examples/distributed-solver/distributed-solver.cpp

```diff
@@ -0,0 +1,263 @@
+    // distributed systems at the moment.
+    using solver = gko::solver::Cg<ValueType>;
+
+    // @sect3{Initialization and User Input Handling}
+    // Since this is an MPI program, we need to initialize and finalize
+    // MPI at the begin and end respectively of our program. This can be easily
+    // done with the following helper construct that uses RAII to automize the
+    // initialization and finalization.
+    const gko::mpi::environment env(argc, argv);
+
+    // Create a MPI communicator wrapper and get the rank.
+    const auto comm = gko::mpi::communicator(MPI_COMM_WORLD);
+    const auto rank = comm.rank();
+
+
+    if (rank == 0) {
+        // Print the ginkgo version information.
+        std::cout << gko::version_info::get() << std::endl;
+    }
+    if (argc == 2 && (std::string(argv[1]) == "--help")) {
+        if (rank == 0) {
+            std::cerr << "Usage: " << argv[0]
+                      << " [executor] [num_grid_points] " << std::endl;
+        }
+        std::exit(-1);
+    }
+
+    ValueType t_init = gko::mpi::get_walltime();
+
+    // User input settings:
+    // - The executor, defaults to reference.
+    // - The number of grid points, defaults to 100.
+    const auto executor_string = argc >= 2 ? argv[1] : "reference";
+    const auto grid_dim =
+        static_cast<gko::size_type>(argc >= 3 ? std::atoi(argv[2]) : 100);
+
+    // Pick requested executor.
+    std::map<std::string, std::function<std::shared_ptr<gko::Executor>()>>
+        exec_map{
+            {"omp", [] { return gko::OmpExecutor::create(); }},
+            {"cuda",
+             [&] {
+                 return gko::CudaExecutor::create(
+                     comm.node_local_rank() %
+                         gko::CudaExecutor::get_num_devices(),
+                     gko::ReferenceExecutor::create(), false,
+                     gko::allocation_mode::device);
+             }},
+            {"hip",
+             [&] {
+                 return gko::HipExecutor::create(
+                     comm.node_local_rank() %
+                         gko::HipExecutor::get_num_devices(),
+                     gko::ReferenceExecutor::create(), true);
+             }},
+            {"dpcpp",
+             [&] {
+                 auto ref = gko::ReferenceExecutor::create();
+                 if (gko::DpcppExecutor::get_num_devices("gpu") > 0) {
+                     return gko::DpcppExecutor::create(
+                         comm.node_local_rank() %
+                             gko::DpcppExecutor::get_num_devices("gpu"),
+                         ref);
+                 } else if (gko::DpcppExecutor::get_num_devices("cpu") > 0) {
+                     return gko::DpcppExecutor::create(
+                         comm.node_local_rank() %
+                             gko::DpcppExecutor::get_num_devices("cpu"),
+                         ref);
+                 } else {
+                     throw std::runtime_error("No suitable DPC++ devices");
```

Comment:
Maybe add something similar for NVIDIA / AMD GPUs?

Reply:
this gets added in #1117, so I won't change anything here